### PR TITLE
Trwalke/ui automation for mobile

### PR DESCRIPTION
**Pull request by trwalke**
Intrgrating Xamarin UI Test framework.
migrating the Azure Keyvault Access classes to the dev branch

Task (https://identitydivision.visualstudio.com/DevEx/_workitems/edit/480655)

### CODE REVIEW

File: adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                new UserQueryParameters
+                {
+                    IsMamUser = false,
+                    IsMfaUser = false,
+                    IsFederatedUser = false
+                });
+
+            controller.Tap("secondPage");
+            controller.Tap("acquireToken");
+            controller.EnterText("i0116", user.Upn, true);
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs:26]: >EnterText [](start = 23, length = 9)

is there a faster way to insert text in a field? e.g. via copy/paste? when watching the automation running this seems to take forever. #Resolved

File: adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            Tap(elementID, isWebElement, defaultSearchTimeout);
+        }
+
+        public void Tap(string elementID, int waitTime, bool isWebElement)
+        {
+            Tap(elementID, isWebElement, new TimeSpan(0, 0, waitTime));
+        }
+
+        public void EnterText(string elementID, string text, bool isWebElement)
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs]: >EnterText [](start = 20, length = 9)

I mentioned this in another place. For some of these text entering elements the automation takes forever. does something better exist in the test infra, to have the text inputs go faster (e.g. via copy/paste). #Resolved

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs (hunk 2 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs]: @henrik-me  The good thing about the Xamarin UI test framework is that it does a copy and paste for text input by default when using "EnterText()". #Resolved

Comment (henrik-me) [reply adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs]: Can the same be used for the desktop tests? #Resolved

**Review by henrik-me — APPROVED** [review]
:shipit:

File: adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs (hunk 2 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/XamarinUITestController.cs]: With the current setup of the desktop tests, slow text input is the only option. However, with newer versions of the selenium framework, it may now be possible. ill investigate. #Resolved

File: adal/devApps/XFormsApp.Droid/XFormsApp.Droid.csproj [hunk 3]
```diff
@@ -31,10 +31,10 @@
     <DefineConstants>DEBUG;TRACE</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <AndroidUseSharedRuntime>True</AndroidUseSharedRuntime>
+    <AndroidUseSharedRuntime>false</AndroidUseSharedRuntime>
```
Comment (jennyf19) [inline adal/devApps/XFormsApp.Droid/XFormsApp.Droid.csproj]: I would do upper case here and line 37...goes w/the others #Resolved

File: adal/devApps/XFormsApp/SecondPage.cs [hunk 4]
```diff
@@ -173,6 +186,7 @@ void LogCallback(LogLevel level, string message, bool containsPii)
             {
                 Device.BeginInvokeOnMainThread(() =>
                 {
+                    this.testResult.Text = string.IsNullOrWhiteSpace(accessToken) ? "Succsess: False" : "Succsess: True";
```
Comment (jennyf19) [inline adal/devApps/XFormsApp/SecondPage.cs]: Check spelling on `Success` #Resolved

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Test.ADAL.NET.UIAutomation
+{
+    public static class CertificateHelper
+    {
+        /// <summary>
+        /// Try and locate a certificate matching the given <paramref name="thumbprint"/> by searching in 
+        /// the in the <see cref="StoreName.My"/> store name for all all available <see cref="StoreLocation"/>s.
```
Comment (jennyf19) [inline adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs]: repeated "in the in the" and "all all available" #Resolved

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        /// <summary>
+        /// The authentication type to use to communicate with the Key Vault.
+        /// </summary>
+        public KeyVaultAuthenticationType AuthType { get; set; }
+
+        /// <summary>
+        /// The ID of the test harness client.
+        /// </summary>
+        /// <remarks>
+        /// This should be configured as as having access to the Key Vault instance specified at <see cref="Url"/>.
```
Comment (jennyf19) [inline adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultConfiguration.cs]: "as as" #Resolved

**Review by jennyf19 — COMMENTED** [review]
this will be great @trwalke - excited to try it out

File: adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+<Project DefaultTargets="Build" ToolsVersion="12.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
     <ProjectGuid>{B19C892E-2628-4CA7-AD27-08D406A3B14B}</ProjectGuid>
     <OutputType>Library</OutputType>
     <RootNamespace>Test.ADAL.NET.UIAutomation</RootNamespace>
     <AssemblyName>Test.ADAL.NET.UIAutomation</AssemblyName>
-    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.1</TargetFrameworkVersion>
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: Why downgrade TargetFramework ? #Pending

File: adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj [hunk 8]
```diff
@@ -32,10 +33,10 @@
     <PackageReference Include="Xamarin.UITest" Version="2.2.4" />
     <PackageReference Include="NUnitTestAdapter" Version="2.1.1" />
     <PackageReference Include="Newtonsoft.Json" Version="6.0.0" />
-    <PackageReference Include="Microsoft.Azure.KeyVault" Version="3.0.0" />
-    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="2.1.1" />
-    <PackageReference Include="Microsoft.Extensions.Configuration.FileExtensions" Version="2.1.1" />
+    <PackageReference Include="Microsoft.Azure.KeyVault" Version="2.0.6" />
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: Why  downgrade these?

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs [hunk 9]
```diff
@@ -54,12 +56,12 @@ public class KeyVaultSecretsProvider
         ///         "clientId": [client ID]
         /// </para>
         /// </remarks>
-        /// <param name="section">Configuration section for Key Vault</param>
         public KeyVaultSecretsProvider()
         {
             _config = new KeyVaultConfiguration();
-            _config.AuthType = KeyVaultAuthenticationType.UserCredential;
+            _config.AuthType = KeyVaultAuthenticationType.ClientCertificate;
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs]: has info on how to place the cert in the right location been added to the "how to setup a new build/test machine"?

File: adal/devApps/XFormsApp.Droid/Properties/AndroidManifest.xml [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" android:installLocation="auto" package="com.Microsoft.XFormsDroid">
```
Comment (bgavrilMS) [inline adal/devApps/XFormsApp.Droid/Properties/AndroidManifest.xml]: Are these changes needed?

File: adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+{
+	public class AppInitializer
+	{
+		public static IApp StartApp(Platform platform)
+		{
+			if (platform == Platform.Android)
+			{
+				return ConfigureApp.Android.InstalledApp("com.Microsoft.XFormsDroid").StartApp();
+
+            }
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs]: Spacing..

File: adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+			return ConfigureApp.iOS.StartApp();
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs]: It'll be safer to add a switch statement over that Platform enum and to return an app for each supported platform and a NotImplemented or NotSupported exception for the other ones.

File: adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using Xamarin.UITest;
+using Xamarin.UITest.Queries;
+
+namespace Test.ADAL.NET.UIAutomation
+{
+	public class AppInitializer
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs]: This class is the very definition of a factory :), consider naming it as such.

Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/AppInitializer.cs]: Please document all public classes and public methods, even for test code.

File: adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+            controller.Tap("idSIButton9", true);
+            controller.EnterText("i0118", ((LabUser)user).GetPassword(), true);
+            controller.Tap("idSIButton9", true);
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs:29]: Do you control these IDs? If so, can you rename them to something more human readable? if not, can you add comments on what each point at? idSIButton9 is not very readable :)

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs [hunk 15]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Test.ADAL.NET.UIAutomation
+{
+    public static class CertificateHelper
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs]: It would help reviewers if you mentioned that you've copied code over from an existing repo, as we wouldn't have to review this again. You can add comments to your own PR. #Resolved

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+        private KeyVaultClient _keyVaultClient;
+
+        private ClientAssertionCertificate _assertionCert;
+
+        private KeyVaultConfiguration _config;
+
+        private readonly string keyVaultClientID = "ebe49c8f-61de-4357-9194-7a786f6402b4";
+
+        private readonly string keyVaultThumbPrint = "440A5BE6C4BE2FF02A0ADBED1AAA43D6CF12E269";
```
Comment (bgavrilMS) [inline adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs]: You might want to run our internal compliance build before commiting, I think it might fail as it will consider there to be credentials.

**Review by bgavrilMS — APPROVED** [review]
Added some minor comments, bug LGTM otherwise. Please fix and commit.

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs (hunk 9 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/KeyVaultSecretsProvider.cs]: Yes

File: adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj (hunk 8 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: @henrik-me newer versions ran into compatibility issues that prevented the build from passing on VSTS

File: adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj (hunk 7 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: Same as other comment
 #Resolved

Comment (henrik-me) [reply adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: I can't find the other comment...

File: adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj (hunk 8 above)
Comment (henrik-me) [reply adal/tests/Test.ADAL.NET.UIAutomation/Test.ADAL.NET.UIAutomation.csproj]: compatibility or missing components on the box?

---
In reply to: [199915776](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1109#discussion_r199915776) [](ancestors = 199915776)

File: adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs (hunk 15 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/Infrastructure/CertificateHelper.cs]: @bgavrilMS good point, ill clearly indicate that next time. thanks

File: adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs (hunk 14 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/CoreTests.cs:29]: The ones that are readable are controllable by me, the others are not. ill add comments.

File: adal/tests/Test.ADAL.NET.UIAutomation/AppFactory.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+		public static IApp StartApp(Platform platform)
+		{
+            switch(platform)
+            {
+                case Platform.Android:
+                    return ConfigureApp.Android.InstalledApp("com.Microsoft.XFormsDroid.ADAL").StartApp();
+                case Platform.iOS:
+                    return ConfigureApp.iOS.StartApp();
+                default:
+                    return ConfigureApp.Android.InstalledApp("com.Microsoft.XFormsDroid.ADAL").StartApp();
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/AppFactory.cs]: > return ConfigureApp.Android.InstalledApp [](start = 19, length = 41)

IMO: this should throw an exception saying platform not supported

File: adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+            this.platform = platform;
+        }
+
+        /// <summary>
+        /// Initializes test controller
+        /// </summary>
+        [SetUp]
+        public void BeforeEachTest()
+        {
+            app = AppFactory.StartApp(platform);
```
Comment (henrik-me) [inline adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs]: >app [](start = 12, length = 3)

are you sure you want to re-create the app with-out setting it in the controller? This makes very little sense to me. Would be great if you can make the code reveal the intent

File: msal/tests/dev apps/XForms/XForms/AcquirePage.xaml [hunk 19]
```diff
@@ -65,14 +65,17 @@
...
         <ScrollView Orientation="Vertical" VerticalOptions="FillAndExpand">
             <StackLayout Orientation="Vertical" VerticalOptions="FillAndExpand">
                 <Frame OutlineColor="Black" Padding="10">
-                    <Label x:Name="acquireResponseLabel" Text="" />
+                    <Label x:Name="acquireResponseLabel" Text=""/>
                 </Frame>
-                <Button  Text="Clear"  Clicked="OnClearClicked" TextColor="Blue" Font="Bold,15"/>
+                <StackLayout Orientation="Horizontal"  HorizontalOptions="FillAndExpand">
+                    <Button  Text="Clear Result"  Clicked="OnClearClicked" TextColor="Blue" Font="Bold,15"/>
+                    <Button  Text="Clear Cache"  Clicked="OnClearCacheClicked" TextColor="Blue" Font="Bold,15" AutomationId="clearCacheButton"/>
```
Comment (henrik-me) [inline msal/tests/dev apps/XForms/XForms/AcquirePage.xaml]: any reason you are not setting AutomationId on all elements?

File: msal/tests/Test.MSAL.NET.UIAutomation/AppFactory.cs [hunk 20]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using Xamarin.UITest;
+using Xamarin.UITest.Queries;
+
+namespace Test.MSAL.NET.UIAutomation
+{
+    /// <summary>
+    /// Initializes the app object that represents the main gateway to interact with the app on the device
+    /// </summary>
+	public class AppFactory
```
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.UIAutomation/AppFactory.cs]: >AppFactory [](start = 14, length = 10)

any way this can be shared between ADAL and MSAL ... e.g. by a shared test core?

seems like we would like to reuse as much as possible including the actual tests. 

thoughts?

Comment (trwalke) [reply msal/tests/Test.MSAL.NET.UIAutomation/AppFactory.cs]: Yes, I can do that

File: msal/tests/dev apps/XForms/XForms/AcquirePage.xaml (hunk 19 above)
Comment (trwalke) [reply msal/tests/dev apps/XForms/XForms/AcquirePage.xaml]: The UI framework does not seem to work well with the tabbed page view that is used for the MSAL app. it can only access the front page thus, I will add IDs for all elements on the front page. We will most likely need to build an application that is dedicated to UI testing and is the same across both libraries.

File: adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs (hunk 18 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/XamarinDroidADALTests.cs]: this is a bug, resolved

File: adal/tests/Test.ADAL.NET.UIAutomation/AppFactory.cs (hunk 17 above)
Comment (trwalke) [reply adal/tests/Test.ADAL.NET.UIAutomation/AppFactory.cs]: @henrik-me I agree

File: core/tests/Test.Microsoft.Identity.Core.UIAutomation/AppFactory.cs [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+        public static IApp StartApp(Platform platform, string targetApp)
+        {
+            switch (platform)
+            {
+                case Platform.Android:
+                    return ConfigureApp.Android.InstalledApp(targetApp).StartApp();
+                case Platform.iOS:
+                    return ConfigureApp.iOS.StartApp();
+                default:
+                    throw new Exception("Platform not supported.");
```
Comment (henrik-me) [inline core/tests/Test.Microsoft.Identity.Core.UIAutomation/AppFactory.cs]: >throw new Exception("Platform not supported."); [](start = 20, length = 47)

Pls. create a specific exception for this. e.g. PlatformNotSupportedException

**Comment by henrik-me** [comment]
Looks really good, great refactorings. Only one minor comment remaining from my side.

**Review by bgavrilMS — APPROVED** [review]
LGTM
